phd-tests/framework/src/artifacts/buildomat.rs

```diff
@@ -0,0 +1,264 @@
+                anyhow::bail!(
+                    "'{c}' is not a valid hexadecimal digit; Git \
+                    commit hashes should consist of the characters \
+                    [0-9, a-f, A-F]"
+                );
+            }
+        }
+
+        Ok(Self(s.to_string()))
+    }
+}
+
+impl fmt::Display for Commit {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(&self.0)
+    }
+}
+
+impl<'de> Deserialize<'de> for Commit {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let s = String::deserialize(deserializer)?;
+        FromStr::from_str(&s).map_err(serde::de::Error::custom)
+    }
+}
+
+impl Series {
+    pub(super) const fn from_static(s: &'static str) -> Self {
+        Self(Cow::Borrowed(s))
+    }
+}
+
+impl fmt::Display for Series {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl BuildomatArtifact {
+    pub(super) fn uri(&self, filename: impl AsRef<Utf8Path>) -> String {
+        let Self {
+            repo: Repo(ref repo),
+            series: Series(ref series),
+            commit: Commit(ref commit),
+            ..
+        } = self;
+        let filename = filename.as_ref();
+        format!("{BASE_URI}/file/{repo}/{series}/{commit}/{filename}")
+    }
+}
+
+impl fmt::Display for BuildomatArtifact {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let Self {
+            repo: Repo(ref repo),
+            series: Series(ref series),
+            commit: Commit(ref commit),
+            ..
+        } = self;
+        write!(f, "Buildomat {repo}/{series}@{commit}")
+    }
+}
+
+impl super::DownloadConfig {
+    pub(super) fn download_buildomat_uri(
+        &self,
+        uri: &str,
+    ) -> anyhow::Result<bytes::Bytes> {
+        tracing::info!(
+            timeout = ?self.timeout,
+            %uri,
+            "Downloading file from Buildomat...",
+        );
+        let client = reqwest::blocking::ClientBuilder::new()
+            .timeout(self.timeout)
+            .build()?;
+        let try_download = || {
+            let request = client
+                .get(uri)
+                .build()
+                // failing to build the request is a permanent (non-retryable)
+                // error, because any retries will use the same URI and request
+                // configuration, so they'd fail as well.
+                .map_err(|e| backoff::Error::permanent(e.into()))?;
+
+            let response = client
+                .execute(request)
+                .map_err(|e| backoff::Error::transient(e.into()))?;
+            if !response.status().is_success() {
+                // when downloading a file from buildomat, we currently retry
+                // all errors, since buildomat returns 500s when an artifact
+                // doesn't exist. hopefully, this will be fixed upstream soon.
```

Comment:
We should file a buildomat tracking issue for this if we don't have one already and link it here.

Reply:
I have a PR that i'm _hoping_ fixes the 500s: https://github.com/oxidecomputer/buildomat/pull/48